### DISCUSSION TITLE
Cached themes - manually clear cache on theme updates

### DISCUSSION THREAD

**Discussion by laskasn** [keycloak/keycloak#8628] [Q&A]
Hello everybody.
We have a custom theme which is packed as a jar and imported as a module (under keycloak modules folder). 
Whenever we have some updates in the theme code, we want to be able to redeploy our theme on an existing keycloak deployment, without having the cache disabled. 
Our theme apart from the static files (js, html, ftl, css, etc} also includes an SPI (java code) to allow fetching some extra data from the keycloak. 
We could hot deploy the jar (which extracts all the jar elements in a new folder) as described in the documentation, but in this case, it doesn't load the SPI(s) at all.
If we replace the jar in the modules folder and restart, it does not refresh the cached theme (js, html, etc).
Currently, we need to always have the theme caching disabled, which is something we do not prefer.
I cannot find any way to have this type of module redeployed without disabling the cache, maybe i am missing something in the documentation.
A dirty hack would be to do (run) something which would clear all cached themes, if such a thing is possible.
Does anyone have a clue on that?

**Comment by laskasn** [comment]
As i see in the code, it seems that there are:

1. a double layer of template caching (one in keycloak and one in Freemarker library) 
2. a theme cache in org.keycloak.theme.DefaultThemeManagerFactory

- Templates caches: The first one is in FreeMarkerUtil, which caches the templates (and is a volatile in-memory map structure) which should clear upon server restarts, while, the second one is the freemarker-2.3.31.jar!/freemarker/cache/TemplateCache.class. Reading the freemarker documentation about the Cache **[here](https://freemarker.apache.org/docs/pgui_config_templateloading.html#autoid_39)** it sais _"Note that some template loaders won't see that a template was changed because of the underlying storage mechanism doesn't support that; for example, class-loader based template loaders may have this problem."_ 
- The theme cache, which is a in-memory map caching the whole theme structure. This is also in-memory and i presume it clears up upon  keycloak node restarts.

@stianst might have a clue about this issue, since as i see he probably implemented the whole keycloak-side theme caching.

It would be great if i could get a clue about the keycloak theme's cache lifecycle:

- when and where it's cached (it's in-memory for the map structures, but where does the freemarker-2.3.31.jar!/freemarker/cache/TemplateCache.class caches the templates?)
- when a cache clear is triggered? (also ftl template cache clear) ?

**Reply by pcolmer** [reply]
This is really causing a problem for me at the moment. I've been working on an updated theme and, for the purposes of testing during development, I've had the cache disabled. If I reconfigure the config to reenable the cache and restart Keycloak, the browser loads the previously cached files and not the latest version of the theme/CSS files.

I can keep caching disabled for now to work around the problem but the fact that something is remaining cached even after a Keycloak restart is problematic ... and this is on a standalone configuration as well.

**Comment by stianst** [comment]
The proper solution to this is most likely to generate a new resource version either automatically if themes have changed, or have a way to specify the resource version. That would result in new URLs for everything, which moves around the browser caching issue.

A better, but more complicated approach would be to have separate resource version for Keycloak built-in stuff, and custom themes, or even go further than that. I'd probably at least initially just do one version for everything, that perhaps separate KC stuff from custom stuff.

**Comment by LatinSuD** [comment]
I found the cache.
It is in /opt/keycloak/standalone/tmp/kc-gzip-cache

There are lots of GZ files there, you just need to delete them for cache to be cleared.
Files are regenerated automatically as they get browsed.

The funny thing is that you don't even need to edit configuration or restart anything. Just edit your theme file, delete the GZ and you're ready to go.

**Reply by ingokuba** [reply]
Thank you! :pray:

If referring to the legacy Keycloak installation with JBoss the correct path is:
```
/opt/jboss/keycloak/standalone/tmp/kc-gzip-cache
```
(was missing `jboss` path part)

**Reply by arulrajnet** [reply]
Adding the following in `standalone.xml` or `standalone-ha.xml` solves the problem for me

```
<theme>
    <staticMaxAge>-1</staticMaxAge>
    <cacheThemes>false</cacheThemes>
    <cacheTemplates>false</cacheTemplates>
    ...
</theme>
```

refer https://keycloakthemes.com/blog/how-to-turn-off-the-keycloak-theme-cache

**Comment by lmorocz** [comment]
FYI. For the Quarkus (19.0.1) version this cache directory is `keycloak/data/tmp/kc-gzip-cache`. 

I had to manually delete this directory after new theme provider jar version deployment. Until then the CSS files remained from the old jar version (although the template changes were ok without the manual data/tmp directory cleaning 🤔).

**Comment by navdeepbansal** [comment]
how can i make reload ftl files ? i have edited some ftl files and cleared the cache at mentioned location but this didnt reflect to keycloak. However on server restart changes appear on Keycloak. Does keycloak has a different caching mechanism for ftl.

`/opt/jboss/keycloak/standalone/tmp/kc-gzip-cache`
